pallets/reward-distribution/src/lib.rs

```diff
@@ -99,5 +161,130 @@ pub mod pallet {
+		let reward_per_block = match RewardPerBlock::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("Reward per block is None");
+				return
+			},
+		};
+
+		if let Err(_) = ext::security::ensure_parachain_status_running::<T>() {
+			return
+		}
+
+		let rewards_adapted_at = match RewardsAdaptedAt::<T>::get() {
+			Some(value) => value,
+			None => {
+				log::warn!("RewardsAdaptedAt is None");
+				return
+			},
+		};
+
+		let mut reward_this_block = reward_per_block;
+
+		//update the reward per block if decay interval passed
+		if let Ok(expired) = ext::security::parachain_block_expired::<T>(
+			rewards_adapted_at,
+			T::DecayInterval::get() - T::BlockNumber::one(),
+		) {
+			if expired {
+				println!("updating rewards per block with decay");
+				let decay_rate = T::DecayRate::get();
+				reward_this_block = (Perquintill::one() - decay_rate).mul_floor(reward_per_block);
+				RewardPerBlock::<T>::set(Some(reward_this_block));
+				RewardsAdaptedAt::<T>::set(Some(height));
+			}
+		} else {
+			log::warn!("Failed to check if the parachain block expired");
+		}
+
+		//TODO how to handle error if on init cannot fail?
+		let _ = Self::distribute_rewards(reward_this_block, T::GetNativeCurrencyId::get());
+	}
+
+	//fetch total stake (all), and calulate total usd stake in percentage across pools
+	//distribute the reward accoridigly
+	fn distribute_rewards(
+		reward_amount: BalanceOf<T>,
+		reward_currency: T::Currency,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		//calculate the total stake across all collateral pools in USD
+		let total_stakes = ext::pooled_rewards::get_total_stake_all_pools::<T>()?;
+
+		let mut total_stake_in_usd = BalanceOf::<T>::default();
+		for (currency_id, stake) in total_stakes.clone().into_iter() {
+			let stake_in_usd = T::OracleApi::currency_to_usd(&stake, &currency_id)?;
+			total_stake_in_usd = total_stake_in_usd.checked_add(&stake_in_usd).unwrap();
+		}
+		//distribute the rewards to each collateral pool
+		let mut percentages_vec = Vec::<(T::Currency, Perquintill)>::new();
+		let mut error_reward_accum = BalanceOf::<T>::zero();
+		for (currency_id, stake) in total_stakes.into_iter() {
+			let stake_in_usd = T::OracleApi::currency_to_usd(&stake, &currency_id)?;
+			let percentage = Perquintill::from_rational(stake_in_usd, total_stake_in_usd);
+			let reward_for_pool = percentage.mul_floor(reward_amount);
+
+			if ext::pooled_rewards::distribute_reward::<T>(
+				&currency_id,
+				&reward_currency,
+				reward_for_pool,
+			)
+			.is_err()
+			{
+				error_reward_accum =
+					error_reward_accum.checked_add(&reward_for_pool).ok_or(Error::<T>::Overflow)?;
+			}
+
+			percentages_vec.push((currency_id, percentage))
+		}
+
+		//we store the calculated percentages which are good as long as
+		//prices are unchanged
+		let bounded_vec = BoundedVec::try_from(percentages_vec)
+			.expect("should not be longer than max currencies");
+		RewardsPercentage::<T>::set(Some(bounded_vec));
```

Comment:
> I left it because I wasn't sure that the price inside one block could change. If it couldn't and the prices change at the beginning only (given the oracles off-chain workers), I thought we could use this percentages and save the first iteration to calculate the amount for the fee distributions.

The price feeding happens as part of a regular extrinsic call. The 'specialty' about this is that this call is done by the off-chain worker but at the end of the day it does not really matter. It's just that every block will have a call to `diaOracleModule::setUpdatedCoinInfos()`, as you can see in eg. [this](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Famplitude-rpc.dwellir.com#/explorer/query/0x90d32c4894e8d47081b947fff8d4dd9b341d6ad5628b50227b28960ef7837e13) block.
But regardless of that, I am not sure how we could use the `RewardsPercentage` in order not to do the first iteration. 

We have to assume that the price of each asset changes every block (since the off-chain worker feeds a new price every block, and it may or may not be the same as in the last block but we can't know).
Thus, even if the values of the `total_stakes` iter did not change between the blocks (which we also don't know), the returned results of any call to `T::OracleApi::currency_to_usd` might change and thus the whole computation has to be re-done.

Or am I missing something? As far as I understand, storing the `RewardsPercentage` on-chain only makes sense if we also add some logic to know if the stakes of the pooled rewards pallet changed or the prices. And this logic would probably cause more overhead than just redoing the calculation each time.

Reply:
Thanks for the explanation, it is more clear now. Indeed, it does not make sense to hold this vector at all. I will remove it.